### ISSUE TITLE
remove the socket file before the listen and change_mode the socket file right after listen

### ISSUE THREAD

**Issue by derek-zhou** [ninenines/ranch#320]
ranch_tcp supports UNIX domain socket via: {local, PATH}. However, it is hard to use in real life because:

* if the file exists, the listen will fail (think supervisor triggered restart)
* the permission to the file is inherited; while in many cases the users want to make it wide open as 0777

I think the best option is to do the remove and change_mode inside `ranch_tcp:listen/1`. I can provide a PR for this.

**Comment by essen** [comment]
If I remember correctly, the file is not automatically removed to prevent mistakenly deleting a file that shouldn't be deleted.

If you embed listeners in your own supervisor, you can easily remove the file before starting the listener.

I'm not sure there are options to set the permissions currently? How would that work? Do you set them after the listener has started?

cc @juhlig

**Comment by derek-zhou** [comment]
Currently, you have to remove the file before the part of the supervision tree is started and change_mode after. However, it is kind of ugly and may not work well in case of a partial crash. It would be nice to have it done automatically, or by passing additional options down together with the `{ip, {local, PATH}}` etc.

**Comment by essen** [comment]
You can handle it automatically by embedding the listener in your supervision tree, in that case you can have your own start_link function that does everything needed.

If you have a way to reproduce a partial crash where this is still not ideal, please provide it.

For non-embedded listeners we can certainly improve the behavior.

**Comment by derek-zhou** [comment]
I am using phoenix, which in turn uses cowboy, which in turn uses ranch. Phoenix can use either a ip socket or an local socket by run time comfiguration; however, ip socket works but local socket does not, due to the reasons above. HTTP over Unix domain socket is non-standard but useful behind a reverse proxy; and in all likelihood the pre remove and post chmod are needed. I still think pushing down this common functionality is the most elegant solution.

**Comment by essen** [comment]
Both Cowboy and Ranch support embedding listeners, and it shouldn't be too difficult to do the same with Phoenix. I am mentioning this solution because it is something you can use now without too much effort.

Let's wait for @juhlig about a more long term solution.

**Comment by juhlig** [comment]
The socket file should already be removed automatically in Ranch 2.0, as of #292.

**Comment by essen** [comment]
Right, thanks.

What about file permissions? How would that work? @derek-zhou where do you think this should be done?

**Comment by derek-zhou** [comment]
thanks @juhlig 
For file permission, I want to pass down the intended permission as an option tuple to ranch and chmod right after the listen. IMHO, this is the cleanest solution.

**Comment by juhlig** [comment]
@derek-zhou yes, just my thinking.

I'll look into it next week I guess. This week I'm on vacation :)

**Comment by juhlig** [comment]
... and @Maria-12648430 is on her no-tech vacation herself for 2 more weeks, so unless @essen steps in, that's earliest ^^;

**Comment by essen** [comment]
We are going to add a callback that will let you do whatever you want to the socket. See #322 

Closing in favor of the PR.

**Comment by essen** [comment]
Please try current master as the PR has been merged.

**Comment by derek-zhou** [comment]
Nice! I will file a issue against cowboy to make use of this callback and chmod 0777 for Unix sockets.

**Comment by essen** [comment]
There's nothing to do in Cowboy.
